### DISCUSSION TITLE
Generic SAML Login Provider

### DISCUSSION THREAD

**Discussion by routerino** [outline/outline#2155] [Ideas]
I've noticed that you're doing work to make generic login providers a thing, and it's fantastic.

However, you're only opening up specific login providers at a time (starting with Microsoft). This seems to go against the idea of having a generic protocol for people to use.

It would make a lot more sense to provide a generic "third party provider" button that instead uses the [saml](https://auth0.com/blog/how-saml-authentication-works/) protocol, which passport [supports](https://github.com/node-saml/passport-saml).

That way rather than having to create a specific use case for *every single provider*, you can just create one. A SAML template that other IdPs can integrate. Most major IdPs can integrate SAML in some form or another, and it vastly reduces the work you have to do to integrate on a per provider basis.

**Comment by tommoor** [comment]
Hi @routerino – generic SAML support is provided in the enterprise edition, I won't be adding it to the open version of the repo. This is one of the places I've chosen to draw the line in order to make Outline a sustainable business.

**Reply by routerino** [reply]
fair enough, gotta monetize somehow

**Reply by routerino** [reply]
> Hi @routerino – generic SAML support is provided in the enterprise edition, I won't be adding it to the open version of the repo. This is one of the places I've chosen to draw the line in order to make Outline a sustainable business.

@tommoor  As a sidebar, could you consider adding [keycloak](https://www.keycloak.org/) as a specific provider? That way the people who really want to disconnect from third party vendors can do so. It's gonna be complicated to do so (you'd have to host both keycloak and outline, neither of which is a trivial process) but it at least provides an option and keycloak is a darn secure auth provider.

The other alternative is local auth, but that's hard to do right.

**Reply by almereyda** [reply]
#2388 has landed now. Could you test and report back if this works with your Keycloak installation?
